#167165072 Build Functionality for Users to Search Articles and Custom filtering

mekzy-o opened the pull request:
#### What does this PR do?

Add functionality to search for articles and custom filtering

#### Description of Task to be completed?

- Users should be able to search for articles by hitting the route `GET -> /api/v1/search/keyword?='bad'&customFilter='author'`
- Users should be able to perform a custom filter on articles by `author's name`, `article tag` and `article title`

*Flowchart for API flow to search article*
![API Flowchart (2)](https://user-images.githubusercontent.com/40548599/62715410-5de7a180-b9f8-11e9-9651-c03a55c0b54f.png)

#### How should this be manually tested?

After cloning this repo, `cd` into it and do the following:

```bash
# fetch this branch
$ git fetch origin  ft-search-articles-167165072

# install dependencies
$ npm install

# run unit tests
$ npm test

# start the development server
$ npm run server
```
Using postman, log in or register an account and hit the get route to search for articles using the `keyword` and `customFilter` query parameter to filter through articles


#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#167165072](https://www.pivotaltracker.com/story/show/167165072)

#### Screenshots (if appropriate)

*User searching for an article by article title*
<img width="1041" alt="Screenshot 2019-08-08 at 9 44 57 PM" src="https://user-images.githubusercontent.com/40548599/62738468-47f2d480-ba2a-11e9-84c4-6f414d7b26b7.png">


*User searching for article by article author*
<img width="1186" alt="Screenshot 2019-08-08 at 9 45 55 PM" src="https://user-images.githubusercontent.com/40548599/62738391-1da11700-ba2a-11e9-9a7d-db82935159b6.png">

*User searching for article by article tag*
<img width="1166" alt="Screenshot 2019-08-08 at 9 46 18 PM" src="https://user-images.githubusercontent.com/40548599/62738509-622cb280-ba2a-11e9-92ce-7b4676b7f3b0.png">

mekzy-o committed:
feat(rating): add user rating feature

Added user rating controller to enable users rate articles
Added integration tests to validate the controller

[Starts #167165062]

mekzy-o committed:
feat(search): add search functionality

Added functionality for users to search articles by Author's name and Article title
Wrote tests to validate functionality
Documented the endpoint using swagger

[Starts #167165072]

mekzy-o committed:
mend

tolulope-od commented on controllers/searchController.js:
```diff
@@ -0,0 +1,131 @@
+              {
+              // tag
+                model: Tag,
+                as: 'tags',
```
Hey Emeka, isn't there a way to search all the related tables and columns in one query? You could do this by combining the sequelize operators

mekzy-o replied:
thanks, done that!

tolulope-od commented on controllers/searchController.js:
```diff
@@ -0,0 +1,131 @@
+      if ((articles.length || tags.length || authors.length) === 0) {
+        return errorResponse(res, 404, { message: 'No Match was found for your request!' });
+      }
+      return successResponse(res, 200, 'matches found', {
```
This key could be difficult to access because of the space, could you please change it to something like `matches`?

PelumiAlesh replied:
+1

mekzy-o replied:
Please check, errorResponse doesn't have a key. Thanks

PelumiAlesh commented on controllers/searchController.js:
```diff
@@ -0,0 +1,122 @@
+    const { keyword } = req.query;
+
+    if (!keyword) {
+      return errorResponse(res, 400, 'Please input a search parameter');
```
Are you passing `Please input a search parameter` as the key here? I think this would be difficult to access in the frontend, you can just use `search` then pass the sentence as a message

OdunayoOkebunmi reviewed (commented):
Emeka, Is your search case sensitive? For instance, will search for 'John' and 'john' produce the same result?

mekzy-o reviewed (commented):
> Emeka, Is your search case sensitive? For instance, will search for 'John' and 'john' produce the same result?

Yes it will

Tunji17 reviewed (approved):
LGTM

PelumiAlesh reviewed (approved):
LGTM

ocranbillions reviewed (approved):
LGTM

mekzy-o committed:
feat(search): reslove merge conflicts with develop

[Starts #167165072]

mekzy-o committed:
feat(search): add custom filter functinality

users should be able filter by tags, authors, and article titles

[Starts #167165072]

mazma1 commented on __tests__/searchRoute.test.js:
```diff
@@ -0,0 +1,81 @@
+        expect(res.status).to.be.eql(200);
+        expect(res.body).to.have.keys('matches');
+        expect(res.body.matches.articles).to.be.a('array');
+        expect(res.body.matches.articles.length).to.be.greaterThan(0);
```
Same correction I made this morning. Don't we have a specific result we can check here?

mekzy-o replied:
it returns an array of objects

mazma1 commented on __tests__/searchRoute.test.js:
```diff
@@ -0,0 +1,81 @@
+        expect(res.status).to.be.eql(200);
+        expect(res.body).to.have.keys('matches');
+        expect(res.body.matches.authors).to.be.a('array');
+        expect(res.body.matches.authors.length).to.be.greaterThan(0);
```
No specific assertion here?

mekzy-o replied:
same here too. You can't tell the specific way it would be fetched when the user types a filter

mazma1 replied:
You should, I assume you are working with a fixed number of test data.

mazma1 commented on __tests__/searchRoute.test.js:
```diff
@@ -0,0 +1,81 @@
+      .end((err, res) => {
+        expect(res.status).to.be.eql(200);
+        expect(res.body).to.have.keys('matches');
+        expect(res.body.matches.tags).to.be.a('array');
```
And here too?

mekzy-o replied:
done. thanks

mekzy-o committed:
Merge branch 'develop' into ft-search-articles-167165072
